### PR TITLE
EVMHost: disallow Byzantium precompiles on earlier EVM versions

### PR DESCRIPTION
**Pull request by axic**

### CODE REVIEW

**Comment by ekpyron** [comment]
So this one would be expected to fail the snark test, right?

**Comment by axic** [comment]
> So this one would be expected to fail the snark test, right?

Yes. Then I'll cherry pick the "fix". I'm kind of hoping some other failure will pop up too, so we have less mysteriously successful tests.

**Comment by ekpyron** [comment]
Seems to fail as expected.

File: test/EVMHost.cpp [hunk 1]
```diff
@@ -115,6 +130,15 @@ void EVMHost::selfdestruct(const evmc::address& _addr, const evmc::address& _ben
 
 evmc::result EVMHost::call(evmc_message const& _message) noexcept
 {
+	if (
+		_message.destination >= 0x0000000000000000000000000000000000000005_address &&
+		_message.destination <= 0x0000000000000000000000000000000000000008_address &&
```
Comment (axic) [inline test/EVMHost.cpp]: I know could have used `> 4 && < 9` but this looked clearer.

File: test/EVMHost.cpp [hunk 2]
```diff
@@ -115,6 +124,17 @@ void EVMHost::selfdestruct(const evmc::address& _addr, const evmc::address& _ben
 
 evmc::result EVMHost::call(evmc_message const& _message) noexcept
 {
+	// Force failure for Byzantium precompiles on pre-Byzantium chains.
+	// NOTE: >=5 && <=8 would be easier to read, but those overloads don't exist in EVMC
+	if (
+		_message.destination > 0x0000000000000000000000000000000000000004_address &&
```
Comment (axic) [inline test/EVMHost.cpp]: Ok I guess can use `!(destination < ...4)` to avoid the new overload.

Comment (ekpyron) [inline test/EVMHost.cpp]:
```suggestion
		0x0000000000000000000000000000000000000004_address < _message.destination &&
```
Then we don't need the operator...

File: test/EVMHost.cpp [hunk 3]
```diff
@@ -115,6 +124,17 @@ void EVMHost::selfdestruct(const evmc::address& _addr, const evmc::address& _ben
...
+		_message.destination < 0x0000000000000000000000000000000000000009_address &&
+		m_evmVersion < langutil::EVMVersion::byzantium()
+	)
+	{
```
Comment (ekpyron) [inline test/EVMHost.cpp]: No brackets needed.

**Review by ekpyron — APPROVED** [review]

**Pull request merged**